command.c

```diff
@@ -5,33 +5,31 @@
 
 #include "nsq.h"
 
-static const int MAX_BUF_SIZE = 128;
-static const int MIN_BUF_SIZE = 64;
-static const int BUF_DELTER   = 2;
+static const int defaultBufSize = 32;
+static const int restBufSize    = 16;
+static const int bufDelter      = 1;
 
-void *nsq_buf_malloc(size_t buf_size, size_t n, size_t l)
+void *nsq_buf_malloc(size_t *buf_size, size_t n, size_t l)
 {
-    if (buf_size - n >= MIN_BUF_SIZE || buf_size - l >= MIN_BUF_SIZE) {
+    int realLeftSize = (int)(*buf_size - n - l);
+
+    if (realLeftSize >= restBufSize) {
         return NULL;
     }
+
+    *buf_size += (restBufSize - realLeftSize) << bufDelter;
```

Comment:
`delter` is not an english word. You should really just do `* 2` here, that's much more clear, and you don't really need to define constants at the top for "double". (Also, the compiler turns `* 2` into `<< 1` pretty reliably.)

But anyway, I don't like this whole function. Here's how I'd remove it entirely:

```diff
--- a/command.c
+++ b/command.c
@@ -5,36 +5,13 @@
 
 #include "nsq.h"
 
-static const int MAX_BUF_SIZE = 128;
-static const int MIN_BUF_SIZE = 64;
-static const int BUF_DELTER   = 2;
-
-void *nsq_buf_malloc(size_t buf_size, size_t n, size_t l)
-{
-    if (buf_size - n >= MIN_BUF_SIZE || buf_size - l >= MIN_BUF_SIZE) {
-        return NULL;
-    }
-    
-    void *buf = NULL;
-    while (1) {
-        if (buf_size - n < MIN_BUF_SIZE || buf_size - l < MIN_BUF_SIZE) {
-            buf_size *= BUF_DELTER;
-            continue;
-        }
-        buf = malloc(buf_size * sizeof(char));
-        assert(NULL != buf);
-        break;
-    }
-    
-    return buf;
-}
+static const int INIT_BUF_SIZE = 128;
+static const int MIN_BUF_FREE = 64;
 
 void nsq_buffer_add(nsqBuf *buf, const char *name, const nsqCmdParams params[], size_t psize, const char *body, const size_t body_length)
 {
-    size_t buf_size = MAX_BUF_SIZE;
+    size_t buf_size = INIT_BUF_SIZE;
     char *b = malloc(buf_size * sizeof(char));
-    char *nb = NULL;
-    assert(NULL != b);
     size_t n = 0;
     size_t l = 0;
 
@@ -52,13 +29,13 @@ void nsq_buffer_add(nsqBuf *buf, const char *name, const nsqCmdParams params[],
                     l = sprintf(b+n, "%d", *((int *)params[i].v));
                     break;
                 case NSQ_PARAM_TYPE_CHAR:
-                    nb = nsq_buf_malloc(buf_size, n, strlen(params[i].v));
-                    if (NULL != nb) {
-                        memcpy(nb, b, n);
-                        free(b);
-                        b = nb;
-                    }
                     l = strlen((char *)params[i].v);
+                    if (buf_size - n < l + MIN_BUF_FREE) {
+                        while (buf_size - n < l + MIN_BUF_FREE) {
+                            buf_size *= 2;
+                        }
+                        b = realloc(b, buf_size);
+                    }
                     memcpy(b+n, (char *)params[i].v, l);
                     break;
             }
@@ -73,11 +50,9 @@ void nsq_buffer_add(nsqBuf *buf, const char *name, const nsqCmdParams params[],
         memcpy(b+n, &vv, 4);
         n += 4;
 
-        nb = nsq_buf_malloc(buf_size, n, body_length);
-        if (NULL != nb) {
-            memcpy(nb, b, n);
-            free(b);
-            b = nb;
+        if (buf_size - n < body_length + MIN_BUF_FREE) {
+            buf_size = n + body_length;
+            b = realloc(b, buf_size);
         }
         memcpy(b+n, body, body_length);
         n += body_length;
```